test/NuGet.Core.Tests/NuGet.PackageManagement.Test/ResolverGatherTests.cs

```diff
@@ -1571,17 +1571,15 @@ public async Task ResolverGather_PackageNamespace_Fails(string packageNamespaces
             };
 
             // Act
-            HashSet<SourcePackageDependencyInfo> results = await ResolverGather.GatherAsync(context, CancellationToken.None);
-
-            List<SourcePackageDependencyInfo> check = results.OrderBy(e => e.Id).ToList();
+            var exception = await Assert.ThrowsAsync<InvalidOperationException>(
+                () => ResolverGather.GatherAsync(context, CancellationToken.None));
 
             // Assert
             Assert.True(namespacesConfiguration.AreNamespacesEnabled);
             Assert.Null(configuredSources);
-            Assert.Equal(1, check.Count);
 
             // Assert log.
-            Assert.Equal($"Package namespace match not found for package ID '{packageId}' ", testNuGetProjectContext.Logs.Value[0]);
+            Assert.Contains($"Package '{packageId} 1.0.0' is not found in the following primary source(s)", exception.Message);
```

Comment:
Is `InvalidOperationException` always thrown when a package download/installation fails?

Reply:
Looks like yes if it it's failing due being unable to resolve package from given sources, I believe it depends of why download/installation is failing.
While debugging `PackageNamespaceInstall_WrongSource_Fails`(it's in [PR#4188](https://github.com/NuGet/NuGet.Client/pull/4188)) test I came [here ResolverGather.cs#L227](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.PackageManagement/Resolution/ResolverGather.cs#L227):
![image](https://user-images.githubusercontent.com/8766776/130303455-48648878-fc68-4fd5-8746-526a4e78208e.png)

Reply:
Sounds good. Resolving this thread.